client/src/api/block_builder/input_selection/new/requirement/native_tokens.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{cmp::Ordering, collections::HashSet};
+
+use primitive_types::U256;
+
+use super::{InputSelection, OutputInfo, Requirement};
+use crate::{
+    block::output::{NativeToken, NativeTokens, NativeTokensBuilder, Output, TokenScheme},
+    error::Result,
+    secret::types::InputSigningData,
+    Error,
+};
+
+pub(crate) fn get_native_tokens<'a>(outputs: impl Iterator<Item = &'a Output>) -> Result<NativeTokensBuilder> {
+    let mut required_native_tokens = NativeTokensBuilder::new();
+
+    for output in outputs {
+        if let Some(output_native_tokens) = output.native_tokens() {
+            required_native_tokens.add_native_tokens(output_native_tokens.clone())?;
+        }
+    }
+
+    Ok(required_native_tokens)
+}
+
+pub(crate) fn get_minted_and_melted_native_tokens(
+    inputs: &[InputSigningData],
+    outputs: &[OutputInfo],
+) -> Result<(NativeTokensBuilder, NativeTokensBuilder)> {
+    let mut minted_native_tokens = NativeTokensBuilder::new();
+    let mut melted_native_tokens = NativeTokensBuilder::new();
+
+    for output in outputs {
+        if let Output::Foundry(output_foundry) = &output.output {
+            let TokenScheme::Simple(output_foundry_simple_ts) = output_foundry.token_scheme();
+            let mut initial_creation = true;
+
+            for input in inputs {
+                if let Output::Foundry(input_foundry) = &input.output {
+                    let token_id = output_foundry.token_id();
+
+                    if output_foundry.id() == input_foundry.id() {
+                        initial_creation = false;
+                        let TokenScheme::Simple(input_foundry_simple_ts) = input_foundry.token_scheme();
+
+                        match output_foundry_simple_ts
+                            .circulating_supply()
+                            .cmp(&input_foundry_simple_ts.circulating_supply())
+                        {
+                            Ordering::Greater => {
+                                let minted_native_token_amount = output_foundry_simple_ts.circulating_supply()
+                                    - input_foundry_simple_ts.circulating_supply();
+
+                                minted_native_tokens
+                                    .add_native_token(NativeToken::new(token_id, minted_native_token_amount)?)?;
+                            }
+                            Ordering::Less => {
+                                let melted_native_token_amount = input_foundry_simple_ts.circulating_supply()
+                                    - output_foundry_simple_ts.circulating_supply();
+
+                                melted_native_tokens
+                                    .add_native_token(NativeToken::new(token_id, melted_native_token_amount)?)?;
```

Comment:
Should we maybe use the minted and melted fields instead? Should be fine the way it's now, if the outputs are correct, but one could be wrong and then we could return the error at this place here already